compiler_gym/datasets/datasets.py

```diff
@@ -0,0 +1,480 @@
+    expensive process of downloading remote resources and caching them locally.
+
+    Use :meth:`deactivate() <compiler_gym.datasets.Datasets.deactivate>` to mark
+    a dataset as inactive such that it won't be included in set of available
+    datasets, but without removing any files that would require a lengthy
+    re-installation:
+
+        >>> env.datasets.install("github-v0")
+        >>> env.datasets.deactivate("github-v0")
+        >>> for dataset in env.datasets:
+        ...     print(dataset.name)
+        benchmark://cbench-v1
+        benchmark://npb-v0
+        >>> env.datasets["github-v0"].installed
+        True
+
+    The :meth:`activate() <compiler_gym.datasets.Datasets.activate>` method can
+    later be used to mark this dataset as ready for use again, without needing
+    to reinstall anything:
+
+        >>> env.datasets.activate("github-v0")
+        >>> for dataset in env.datasets:
+        ...     print(dataset.name)
+        benchmark://cbench-v1
+        benchmark://github-v0
+        benchmark://npb-v0
+
+    The installation of datasets occurs lazily and automatically, however there
+    are :meth:`install() <compiler_gym.datasets.Datasets.install>` and
+    :meth:`uninstall() <compiler_gym.datasets.Datasets.uninstall>` methods
+    should you wish to perform this manually.
+    """
+
+    def __init__(
+        self,
+        datasets: Iterable[Dataset],
+        random: Optional[np.random.Generator] = None,
+    ):
+        # A mapping from dataset name to the dataset instance.
+        self._datasets: Dict[str, Dataset] = {d.name: d for d in datasets}
+        # The set of activate datasets. All datasets except hidden ones are
+        # initialized as activate.
+        self._active_datasets: Set[str] = set(
+            name for name, dataset in self._datasets.items() if not dataset.hidden
+        )
+        self.random = random or np.random.default_rng()
+
+    def seed(self, seed: Optional[int] = None) -> None:
+        """Set the random state.
+
+        Setting a random state will fix the order that
+        :meth:`datasets.benchmark() <compiler_gym.datasets.Datasets.benchmark>`
+        returns benchmarks when called without arguments.
+
+        Calling this method recursively calls :meth:`seed()
+        <compiler_gym.datasets.Dataset.seed>` on all member datasets.
+
+        :param seed: An optional seed value.
+        """
+        self.random = np.random.default_rng(seed)
+        for dataset in self._datasets.values():
+            dataset.seed(seed)
+
+    def datasets(
+        self, inactive: bool = False, hidden: bool = False
+    ) -> Iterable[Dataset]:
+        """Enumerate the datasets.
+
+        Dataset order is consistent across runs.
+
+        :param inactive: If :code:`False` (the default), only datasets that are
+            active will be returned.
+
+        :param hidden: If :code:`False` (the default), only datasets whose
```

Comment:
Why is this needed?

Reply:
Hidden datasets are used so that we can deprecate things gracefully. E.g.

```py
>>> list(env.datasets.dataset())
benchmark://npb-v0
benchmark://cbench-v2
>>> list(env.datasets.dataset(hidden=True))
benchmark://npb-v0
benchmark://cbench-v0
benchmark://cbench-v1
benchmark://cbench-v2
>>> env.datasets.benchmark("benchmark://cbench-v0/crc32")
# works, but will raise a deprecation warning
```

Reply:
Ah, I like that purpose :-)
<pedant-warning>
Consider renaming it to `include_deprecated`. First deprecated makes the purpose of the hidding clear - and will suggest to people that they should not rely on it sticking around. Second with just the name being `hidden`, I would guess it would return the hidden benchmarks, not the unhidden ones.

Reply:
Good idea, will update #200 with that